src/Language/PureScript/Sugar/Names.hs

```diff
@@ -376,54 +394,79 @@ renameInModule imports (Module modSS coms mn decls exps) =
   updateValueName :: Qualified Ident -> SourceSpan -> m (Qualified Ident)
   updateValueName = update (importedValues imports) IdentName
 
+  updateValueName' :: Qualified Ident -> SourceSpan -> m (Maybe (Qualified Ident))
+  updateValueName' = update' (importedValues imports) IdentName
+
   updateValueOpName
     :: Qualified (OpName 'ValueOpName)
     -> SourceSpan
     -> m (Qualified (OpName 'ValueOpName))
   updateValueOpName = update (importedValueOps imports) ValOpName
 
+  updateValueOpName'
+    :: Qualified (OpName 'ValueOpName)
+    -> SourceSpan
+    -> m (Maybe (Qualified (OpName 'ValueOpName)))
+  updateValueOpName' = update' (importedValueOps imports) ValOpName
+
   updateKindName
     :: Qualified (ProperName 'KindName)
     -> SourceSpan
     -> m (Qualified (ProperName 'KindName))
   updateKindName = update (importedKinds imports) KiName
 
   -- Update names so unqualified references become qualified, and locally
-  -- qualified references are replaced with their canoncial qualified names
-  -- (e.g. M.Map -> Data.Map.Map).
+  -- qualified references are replaced with their canoncial qualified names.
+  -- Throws UnknownName error if a name can't be qualified.
   update
     :: (Ord a)
     => M.Map (Qualified a) [ImportRecord a]
     -> (a -> Name)
     -> Qualified a
     -> SourceSpan
     -> m (Qualified a)
-  update imps toName qname@(Qualified mn' name) pos = warnAndRethrowWithPosition pos $
-    case (M.lookup qname imps, mn') of
+  update imps toName qname@(Qualified mn' _) pos = warnAndRethrowWithPosition pos $ do
+    mbQualified <- update' imps toName qname pos
+    maybe reportError pure mbQualified
+    where
+    reportError =
+      case mn' of
 
+        -- If the name wasn't found in our imports but was qualified then we need
+        -- to check whether it's a failed import from a "pseudo" module (created
+        -- by qualified importing). If that's not the case, then we just need to
+        -- check it refers to a symbol in another module.
+        Just mn'' ->
+          if mn'' `S.member` importedQualModules imports || mn'' `S.member` importedModules imports
+          then throwUnknown
+          else throwError . errorMessage . UnknownName . Left . Qualified Nothing $ ModName mn''
+
+        Nothing -> throwUnknown
+
+    throwUnknown = throwError . errorMessage . UnknownName . Left . fmap toName $ qname
+
+  -- Update names so unqualified references become qualified, and locally
+  -- qualified references are replaced with their canoncial qualified names.
+  -- Returns Nothing if name can't be qualified, so that errors can be thrown
+  -- or substitutions can be made.
+  -- (e.g. M.Map -> Data.Map.Map).
+  update'
+    :: (Ord a)
+    => M.Map (Qualified a) [ImportRecord a]
+    -> (a -> Name)
+    -> Qualified a
+    -> SourceSpan
+    -> m (Maybe (Qualified a))
```

Comment:
Assuming it makes sense to have the `UnknownValue` and `UnknownName` constructors use `Name` instead of `Text` for unknown names, presumably we could return an `Either Name (Qualified a)` here so that in the case that the name is not known, we immediately have access to a `Name` to use in the error? Then we don't need to supply the appropriate `Name` constructor more than once.

Reply:
I made `UnknownValue` and `UnknownName` take a `Qualified Name`, so I'm not sure this if this applies anymore.  Let me know if you had something else in mind

Reply:
Yes, this still applies; at the moment, we have to supply a `toName` function in two places: once as an argument to `update'` and once as an argument to `updateOrReplaceValue`. If `update'` returned an `Either Name (Qualified a)` then we wouldn't need to have `updateOrReplaceValue` take a `mkName` parameter, because the `Name` we want is right there already.

Reply:
Ah yep, you're right, I missed the point at first. I believe I've addressed this here https://github.com/purescript/purescript/pull/3619/commits/c8abc2eb8e812eb029a68a5622cb2fa369cdb262